### ISSUE TITLE
Ignore self-usage in functions when determing whether it is used

### ISSUE THREAD

**Issue by bjorn** [mpeterv/luacheck#50]
This function will never be reported as unused, because it is referring to itself:

```lua
local function foo()
    return foo
end
```

This occurs commonly in case of recursion, so it could be worth to ignore these self-references when counting usages.

**Comment by mpeterv** [comment]
Good idea, it can be applied to mutually recursive functions, too.

**Comment by mpeterv** [comment]
Implemented on master branch:

```
Checking test.lua                                 1 warning

    test.lua:1:16: unused recursive function foo

Total: 1 warning / 0 errors in 1 file
```

Also for mutually recursive functions:

```lua
local g

local function f()
   return g()
end

function g()
   return f()
end
```

```
Checking test.lua                                 2 warnings

    test.lua:3:16: unused mutually recursive function f
    test.lua:7:10: unused mutually recursive function g

Total: 2 warnings / 0 errors in 1 file
```

**Comment by bjorn** [comment]
That's really nice that you even take into account mutual recursive functions!

I'm starting to be scared of writing Lua code without luacheck. It prevents so many errors from showing up at runtime and it really helps making sure you're cleaning up any "garbage" while refactoring. Thanks a lot for this tool!

**Comment by mpeterv** [comment]
No problem, thanks for ideas! 0.14.0 released.
